File: credit.c
#include<cs50.h>
#include<stdio.h>
#include<math.h>

int main(void)
{
    long n = get_long("Number: ");    //get the number
    int i = 0, r, o = 0, e = 0;
    long c;
    c = n;                            //copy the number to a seperate variable
    while (c > 0)                     //count the number of digits using while loop
    {
        c /= 10;
        i++;
    }                          
    if (i == 13 || i == 15 || i == 16)        //check for valid number of digits
    {
        r = round(n / 1000000000000);
        while (n > 0)
        {
            o += (n % 10);                //take the sum of digits in the second-to-first position
            n = round(n / 10);
            e += ((((n % 10) * 2) % 10) + (round(((n % 10) * 2) / 
                                                 10))); //multiply the digits in the second-to-last position by two and break them into two separate digits if the product is more than 9, then add all the digits
            n = round(n / 10);
        }
        if ((o + e) % 10 == 0)   //check whether the sum of the sum holds 0 in the last position
        {
            if ((i == 13 && r == 4) || (i == 16 && (r / 1000) == 4))  //if given condition satisfied print VISA
            {
                printf("VISA\n");
            }
            else if (i == 15 && (r / 10 == 34 || r / 10 == 37))  //if given condition satisfied print AMEX
            {
                printf("AMEX\n");
            }
            else if (i == 16 && (r / 100 == 51 || r / 100 == 52 || r / 100 == 53 || r / 100 == 54
                                 || r / 100 == 55)) //if given condition satisfied print MASTERCARD
            {
                printf("MASTERCARD\n");
            }
            else             // if none of the above condition satisfied then print INVALID
            {
                printf("INVALID\n");
            }
        }
        else        //if the initial condition ((o + e)%10 == 0) not satisfied print INVALID
        {
            printf("INVALID\n");
        }
    }
    else        //if the initial condition (i == 13 || i == 15 || i == 16) not satisfied, print INVALID
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
from cs50 import get_int

n = get_int("Number: ")
i = 0
o = 0
e = 0
# copy the value to a new variable
c = n
# get the number of digits
while c > 0:
    c = int(c / 10)
    i += 1
# check whether the length is valid
if i in (13, 15, 16):
    r = int(n / 1000000000000)
    # using LUHN's Algorithm
    while n > 0:
        o += n % 10
        n = int(n / 10)
        e += (((n % 10) * 2) % 10) + int(((n % 10) * 2) / 10)
        n = int(n / 10)
    # if the last dight of the final sum of luhn's algo is 0
    if (o + e) % 10 == 0:
        if (i == 13 and r == 4) or (i == 16 and int(r / 1000) == 4):
            print("VISA")
        elif i == 15 and (int(r / 10) in (34, 37)):
            print("AMEX")
        elif i == 16 and (int(r / 100) in (51, 52, 53, 54, 55)):
            print("MASTERCARD")
        # if not satisfy for visa, amex , mastercard then invalid for this context
        else:
            print("INVALID")

    # if the last digit is NOT zero
    else:
        print("INVALID")
# if the length is not valid
else:
    print("INVALID")